specs/src/recovery.md

```diff
@@ -0,0 +1,525 @@
+the send queue in a FIFO fashion as `Data`. This ensures that a haves take
+longer to be delivered to a congested connection, making it more likely peers
+will recieve that same `Have` message first from a different peer. 
+
+- **Nodes SHOULD send `Have` and `Data` messages from a FIFO queue**
+- **Nodes SHOULD request `Data` from the first peer that sends them a `Have`
+  message**
+
+Lastly, there's still a chance that a peer could already have all `Data`
+included in the `Commitment`, recreate the `Have` messages and send them to many
+nodes in the network extremely quickly. Without additional rules, this would
+cause that node to become a bottleneck in distribution, therefore we add a new
+rule to ensure this doesn't occur.
+
+- **Nodes MUST have fewer than `PerPeerConcurrentRequestCap` concurrent requests
+  to a single peer**
+
+If this value is set to 2 for example, then there will be a max of two
+concurrent requests open to each peer. After the first request is fullfilled,
+then the second is sent.
+
+The above scenario describes a node that is honest and has all of the data, however there are also scenarios where the validator could be malicious which is handled by this mechanism. See the security section for discussing how `PerPeerConcurrentRequestCap` is calculated using the block size, number of potential faulty validators, and the amount of parity data added to the block.
+
+## Security
+
+As discussed in the intro, the above spec is not yet secure. We have to combine
+2 different mechanisms, parity data and a validator overlay network. Before we
+go over how those mechanisms defend against attacks, we first walk through the
+attacks.
+
+Currently, >1/3 voting power could halt the network, therefore any new
+broadcasting mechanism must be held to same standard. Meaning that if 2/3 of the
+voting power is honest, there should not be attacks that can break the block
+propagation mechanism either.
+
+The only assumptions that we can make are:
+
+**- 1) Nodes are not eclipse attacked** **- 2) 2/3 of the validators are
+participating as described in this spec**
+
+The first assumption means that there exists *a* path of honest nodes connecting
+the broadcaster and every other honest node. The second is self explanatory.
+
+Recall the constraints for recovery:
+
+>- **100% of the Block data MUST be delivered to >2/3 of the validators**
+>- **The time to distribute that data MUST be <= the `ProposalTimeout`**
+
+While pull based gossip and broadcast trees are incredibly powerful tools for
+increasing throughput, it's easy to derive attacks that fail these constraints
+even with the assumptions. First we discuss those attacks, and then we will
+discuss strategies for mitigation.
+
+### Denial of Service
+
+The purpose of the strict diconnection rules are to prevent all scenarios where
+an attacker can flood the network with valid messages. If invalid messages are
```

Comment:
Shouldn't this be "flood the network with valid and invalid messages"?

valid messages cases:
- Nodes MUST disconnect from peers that send Data messages that have not been requested
- Nodes MUST disconnect from peers that send more than one of the same Have or Want message

invalid messages case:

- Nodes MUST disconnect from peers that send invalid messages